rfcs/0001-fobnail-architecture-change.md

```diff
@@ -0,0 +1,143 @@
+# Summary
+
+Change Fobnail architecture - instead of having server on the attester and
+client on Fobnail Token, server would be located on Fobnail Token and client on
+Attester.
+
+# Motivation
+
+The current architecture severely limits Fobnail Token configuration abilities
+due to Fobnail Token being a client. Fobnail Token has hardcoded paths for
+performing provisioning and attestation. We are not capable of sending requests
+to Fobnail Token which severely limits the amount configuration possible, e.g.
+an once provisioned platform can not be unprovisioned without resetting Fobnail
+Token into factory state. Fobnail Token configuration cannot be changed without
+resetting and going once again through the provisioning process.
+
+It is also not possible to expose any services from Fobnail Token to Attester,
+such as access to symmetric cryptographic keys or performing cryptographic
+operations on behalf of attester in case of asymmetric cryptography. This is a
+must for features like disk encryption or Fobnail-based authentication to
+services, like VPN.
+
+Switching to a server based Fobnail Token allows to expose services for
+accessing cryptographic keys and performing administrative tasks.
+
+# Architecture change proposal
+
+## Fobnail Token provisioning
+
+Currently, if in unprovisioned state, Fobnail Token initiates provisioning by
+requesting PO certificate chain from Platform Owner, then it verifies PO
+certificate chain against a trusted certificate embedded into Fobnail firmware.
+If verification passes, Fobnail generates Identity/Encryption keypair and
+requests PO to generate certificate from this key. Retrieval of valid
+Identity/Encryption certificate finalizes provisioning process.
+
+Instead we propose the following protocol:
+1. Unprovisioned Fobnail token awaits for server to commence provisioning by
+   sending it's certificate chain.
+2. Fobnail verifies certificate chain against trusted certificate embedded into
+   its firmware. If verification is succesful, Fobnail generates
+   Identity/Encryption keypair and responds with Certificate Signing Request.
+   Otherwise an error is returned.
+3. At this point Platform Owner configures Fobnail, like default appraisal
```

Comment:
I have no idea how point 2 relates to point 3. Why point 3 have to happen before point 2? Maybe refering to exisiting documentation would be helpful to understand the context.

Reply:
@pietrushnic I don't know why you think point 3 have to happen before point 2. It have to happen after.

Reply:
@arturkow2000 I don't think that I just ask, what is the relation between two so they can be one after another and not other way. Since I see no connection between 2 and 3 based on description I challenge it by saying why to not revert those.

Reply:
Step 1 and 2 are closely related and should happen one after another as step 2 verifies data sent in step 1 (certchain) and sends response (either CSR or error), so no swapping should be done.

But, nothing prevent from doing step 3 before step 1, unless Fobnail Token needs to know chain in advance to configure something. For default appraisal policy we don't need certifate so we may do this before.

Reply:
@arturkow2000 so essentially there is no dependency between steps 1/2 and 3. No we have to scenarios either we do configuration either we don't, no matter the case in future we may want to change policy. Shouldn't we have separate section about policies make/remove/change and link it here? TBH if this is step is not mandatory, and we should be able to change it later we should make it optional or remove it to simplify things and point user to the place where policy management is documented.

Reply:
@krystian-hebel if that is the case we should include that information that steps maybe related. @arturkow2000 it looks like based on @krystian-hebel comment it should not be removed, but improved.

Reply:
@krystian-hebel These APIs are meant for one time token provisioning and were ment to be locked after provisioning is complete. It could be changed to allow reconfiguration of token, but then we shouldn't expect to finish configuration by sending certificate as we may not want to change it.

Reply:
Even with one-time provisioning, Token must evaluate that whoever communicates with it is in fact a valid PO. It shouldn't do any configuration (e.g. change default appraisal policy) before that. Even with assumption that provisioning is done in safe environment, data sent to Token should be signed to prove that PO has access to private keys, and not only certificate chain (which may and probably should be publicly available).

To sum up, chain verification must be first, CSR generation must be last, whatever is or will be in between doesn't have to be in strict order. Current version of RFC describes it this way so I'm marking this as resolved.